src/game.cpp

```diff
@@ -11602,14 +11602,78 @@ void game::read()
     auto filter = []( const item &it ) {
         return it.is_book();
     };
-    int pos = inv_for_filter( _("Read:"), filter );
 
-    if (pos == INT_MIN) {
+    auto item_loc = inv_map_splice( filter, _("Read:") );
+
+    const int inv_pos = item_loc.get_inventory_position();
+
+    if( inv_pos != INT_MIN ) {
+        u.reading_from_inventory = true;
+        draw();
+        u.read(inv_pos);
+        return;
+    }
+
+    u.reading_from_inventory = false;
+
+    item *it = item_loc.get_item();
+    if( it == nullptr ) {
         add_msg(_("Never mind."));
         return;
     }
+
+
+    // check if in vehicle as item cannot be in inventory since the above check "inv_pos != INT_MIN" already handle item in inventory
+    // at the moment, only item_location::item_on_person() will set values returned for get_inventory_position(). All others will return INT_MIN
+    bool item_on_veh = false;
```

Comment:
That block is quite impressive, buts it's way to complicated for the simple task of picking up a single item. Just get a reference to the item from the `item_location`, add it using `player`/`Character` functions (`i_add`), and remove the item from the map (via the `item_location`).

You may check with `Character::can_pickVolume` / `Character::can_pickWeight` before adding the item. But there is no point in routing this through `Pickup::do_pickup` if you have to write half a page of code just to determine the arguments for that function.